File: spheres.erl
#!/usr/bin/env escript

%% Main program
main(_) ->
  Iterations = 1000,
  Results = run(Iterations),
  printResults(Results, Iterations).

%%%%%%%%%%%
%%% Bag %%%
%%%%%%%%%%%

%% Create a representation for a bag of black and white balls.
createBag(NBlack, NWhite) ->
  Bag = lists:flatmap(fun({X, N}) -> lists:duplicate(N, X) end, [{b, NBlack}, {w, NWhite}]),
  shuffle(Bag).

%% Draw 2 elements recursively from a bag and return the last remaining element.
recursiveDraw(Bag) ->
  Len = length(Bag),
  if Len > 1 ->

    {Selection, Remaining} = drawTwo(Bag),
    NewBag = if Selection == [b, b] -> Bag;
                Selection == [w, w] -> Remaining;
                Selection == [b, w] -> lists:merge([w], Remaining);
                Selection == [w, b] -> lists:merge([w], Remaining);
                true -> printError('Unknown selection', Selection)
             end,
    recursiveDraw(NewBag);

    Len =:= 1 -> Bag;

    true -> printError('Recursive Draw Error')
  end.

%% Retrieve a single element from a bag and return the element and remainder.
drawOne(Bag) ->
  Len = length(Bag),
  I = random:uniform(Len),
  Elem = lists:nth(I, Bag),
  {Elem, lists:delete(Elem, Bag)}.

%% Retrieve two elements from a bag.
drawTwo(Bag) ->
  {Elem1, Draw1} = drawOne(Bag),
  {Elem2, Remaining} = drawOne(Draw1),
  {[Elem1, Elem2], Remaining}.

%%%%%%%%%%%%%%%%%%
%%% Simulation %%%
%%%%%%%%%%%%%%%%%%

%% Run the simulation the given number of times and return all the final elements.
run(Iterations) ->
  run([], Iterations).
run(PreviousResults, RemainingIterations) ->
  NBlack = 25,
  NWhite = 25,
  Bag = createBag(NBlack, NWhite),
  if RemainingIterations > 0 ->
    NewResult = recursiveDraw(Bag),
    NewResults = lists:merge(NewResult, PreviousResults),
    run(NewResults, RemainingIterations - 1);

    RemainingIterations =:= 0 -> PreviousResults;
    true -> printError('Invalid iteration count', RemainingIterations)
  end.

%% Print the percent likelihood of each ball type being the remaining ball.
printResults(Results, Iterations) ->
  BCount = count(b, Results),
  WCount = count(w, Results),
  BResult = resultString('BLACK', BCount, Iterations),
  WResult = resultString('WHITE', WCount, Iterations),
  io:format(BResult),
  io:format(WResult).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% Convenience Functions %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Print an error with an optional extra argument.
printError(Name, X) ->
  io:format("Error: ~s ~p\n", [Name, X]).
printError(Name) ->
  io:format("Error: ~p\n", [Name]).

%% Randomly shuffle a list
shuffle(L) ->
  random:seed(now()),
  [X||{_,X} <- lists:sort([ {random:uniform(), N} || N <- L])].

%% Return the number of occurrences in a list.
count(Item, L) ->
  C = lists:filter(fun(X) -> X == Item end, L),
  length(C).

%% Build a formatted result string.
resultString(Name, Count, Iterations) ->
  Percentage = (Count / Iterations) * 100,
  io_lib:format("~s: ~.2.0f% [~w / ~w]~n", [Name, Percentage, Count, Iterations]).

File: spheres.py
#!/usr/bin/env python
import random
from collections import Counter


BLACK = 'b'
WHITE = 'w'


class Bag:

    """ A bag of black and white balls """

    def __init__(self, nBlack, nWhite):
        self.contents = [BLACK] * nBlack + [WHITE] * nWhite

    def draw(self):
        """ Remove and return at least 2 balls """
        k = 2 if len(self.contents) >= 2 else 1  # for simplicity
        sample = random.sample(self.contents, k)
        [self.contents.remove(ball) for ball in sample]
        return sample

    def returnBall(self, ball):
        """ Return a ball to the bag """
        self.contents.append(ball)

    def returnBalls(self, balls):
        """ Return multiple balls to the bag """
        [self.returnBall(ball) for ball in balls]


class Simulation:

    """ A simulation for sequentially removing/replacing balls from a bag """

    def __init__(self, iterations, nBlack=25, nWhite=25):
        self.iterations = iterations
        self.nBlack = nBlack
        self.nWhite = nWhite

    def run(self):
        """
        Run the simulation for the given number of iterations
        """
        self.results = []
        for x in xrange(self.iterations):
            bag = Bag(self.nBlack, self.nWhite)
            balls = bag.draw()
            while len(balls) == 2:
                if balls == [BLACK] * 2:
                    bag.returnBalls(balls)
                # elif balls == [WHITE] * 2:
                    # Do nothing!
                elif all(x in balls for x in [BLACK, WHITE]):
                    bag.returnBall(WHITE)
                balls = bag.draw()
            self.results.append(balls[0])
        self.printResults()

    def printResults(self):
        """
        Print the percent likelihood of each ball type being the remaining ball
        """
        c = Counter(self.results)

        def result(name, count):
            return name + ': ' + \
                "{:.2f}".format((count / self.iterations) * 100) + '%' + \
                ' [' + str(count) + '/' + str(self.iterations) + ']'

        print(result('BLACK', c[BLACK]))
        print(result('WHITE', c[WHITE]))


if __name__ == "__main__":
    simulation = Simulation(1000)
    simulation.run()
